src/impl/KokkosComm_types.hpp

```diff
@@ -21,50 +21,88 @@
 #include <Kokkos_Core.hpp>
 
 namespace KokkosComm::Impl {
-template <typename Scalar>
+template <typename T>
 MPI_Datatype mpi_type() {
-  static_assert(std::is_void_v<Scalar>, "mpi_type not implemented");
-  return MPI_CHAR;  // unreachable
-};
+  if constexpr (std::is_same_v<T, std::byte>)
```

Comment:
Is `T` guaranteed to always be an unqualified type? Should we decay `T` to strip away potential const/volatile? Otherwise the logic won't work for `mpi_type<const double>`.

```suggestion
  using scalar_type = std::decay_t<T>;
  if constexpr (std::is_same_v<scalar_type, std::byte>)
```

Reply:
Also, I don't know the kokkos coding rules. Some projects require braces around *all* blocks. Maybe that's not an issue here.

Reply:
> Is `T` guaranteed to always be an unqualified type? Should we decay `T` to strip away potential const/volatile? Otherwise the logic won't work for `mpi_type<const double>`.

This function is meant to be called with T being a `View::value_type`, so I guess the question is whether `View<const double*>::value_type` is `double` or `const double`... But adding such a safety net can't hurt I guess. Any particular reason for using `std::decay_t` over `std::remove_cv_t`?

Reply:
> Also, I don't know the kokkos coding rules. Some projects require braces around _all_ blocks. Maybe that's not an issue here.

I don't know either. @cwpearson what do you prefer?

Reply:
Yes. This code will work with `std::decay_t` but not with `std::remove_cv_t` because :

```C
template<typename T>
void send(T&& val) {
  MPI_Send(&val, 1, mpi_type<T>(), ...);
}

void foo() {
  int i;
  send(i);
}
```

`T` in this case will be `int&`. That's probably not going to happen with `View::value_type` but it might be used for something else in the future.

Reply:
> I don't know either. @cwpearson what do you prefer?

This is something that is easily enforced with `clang-tidfy` when we decide that we care, see https://clang.llvm.org/extra/clang-tidy/checks/readability/braces-around-statements.html.

Reply:
@devreal Maybe `std::remove_cvref_t` instead then, but `std::decay_t` also turn `T[]` into `T*` (and function types into function pointers too) and I don't think that we want this. What do you think?

Reply:
Pointers are caught by the last branch (they're obviously not `void` so we get a `static_assert`. Same for functions.

Reply:
You have a point, it does not matter indeed. Let's go with `decay` then, it's less typing 😅